snputils/snp/io/read/pgen.py

```diff
@@ -102,48 +102,83 @@ def open_textfile(filename):
             pvar_header_line_num = 0
             with open_textfile(pvar_filename) as file:
                 for line_num, line in enumerate(file):
-                    if line.startswith("#CHROM"):
+                    if line.startswith("##"):  # Metadata
+                        continue
+                    elif line.startswith("#CHROM"):  # Header
                         pvar_header_line_num = line_num
+                        header = line.strip().split()
                         break
-                else:  # if no break
-                    pvar_has_header = False
+                    elif not line.startswith("#"):  # If no header, look at line 1
+                        pvar_has_header = False
+                        cols_in_pvar = len(line.strip().split())
+                        if cols_in_pvar == 5:
+                            header = ["#CHROM", "ID", "POS", "ALT", "REF"]
+                        elif cols_in_pvar == 6:
+                            header = ["#CHROM", "ID", "CM", "POS", "ALT", "REF"]
+                        else:
+                            raise ValueError(
+                                f"{pvar_filename} is not a valid pvar file."
+                            )
+                        break
+                    
+            def lazy_read(filename: str, **kwargs) -> pl.LazyFrame:
+                """
+                Simple reader function needed due to lack of support for scanning zstd files in polars.
+                
+                Args:
+                    filename (str): pvar file, either .pvar or .pvar.zst
+                    **kwargs: CSV arguments for polars
+                
+                Returns:
+                    pl.LazyFrame
+                """
+                if filename.endswith('.zst'):
+                    return pl.read_csv(filename, **kwargs).lazy()
+                else:
+                    return pl.scan_csv(filename, **kwargs)
 
-            pvar = pl.scan_csv(
+            pvar = lazy_read(
```

Comment:
Hmm, the scan_csv issue with the zst file is weird because it seems that scan_csv does support reading compressed files (https://github.com/pola-rs/polars/issues/7287), and looking at the code from the PR solving that issue (https://github.com/pola-rs/polars/pull/17841) it seems that support for zst format is indeed included. I'll investigate a bit if the error that appeared in your previous commit (polars.exceptions.ShapeError: The length of the new names list should be equal to or less than the original column length) can be solved so we can use scan_csv for everything, as I guess that .pvar.zst would be the ones that benefit more from not using read_csv, since if they're compressed it's probably because they're large.

Reply:
The issue I had was not in regards to using the `scan_csv()` function for .zst files, but instead the ability to define the header. Compatibility with pvar files that do not have a header would break with that shape error. It just doesn't seem that scanning compressed files has a mature suite of params, unfortunately. And either way, it must be read into memory fully no matter what, so the changes I propose here are merely to avoid the error.

Reply:
Ah, I see. Thanks!

Reply:
No problem, feel free to review further. I can make changes if you see anything else